packages/flutter/lib/src/services/deferred_component.dart

```diff
@@ -0,0 +1,157 @@
+
+import 'dart:async';
+
+import 'package:flutter/foundation.dart';
+
+import 'system_channels.dart';
+
+/// Default Google Play Store implementation states that reflect the current
+/// installation status of a deferred component module.
+///
+/// Returned by [DeferredComponent.getDeferredComponentInstallState].
+///
+/// These states may differ in custom deferred component implementations.
+///
+/// Due to the asynchronous nature of platform channels, these states are
+/// informational and should not be depended on to determine if a dynamic
+/// feature is ready to use or not. Readiness should be determined by
+/// completion of the future returned by
+/// [DeferredComponent.installDeferredComponent] or `loadLibrary()`.
+///
+/// These states are an extension of the states in Android's
+/// https://developer.android.com/reference/com/google/android/play/core/splitinstall/model/SplitInstallSessionStatus
+///
+/// Typical state flow begins as `unknown`, and transitions into `requested`
+/// when the installation is begun. When the request is processed, the
+/// state changes to `downloading` and finally `installed`. Modules previously
+/// installed but not loaded in this session will be in the
+/// `installedPendingLoad` state, which indicates that the installation request
+/// call (either `loadLibrary()` or [DeferredComponent.installDeferredComponent]) should
+/// be repeated to trigger the loading process.
+enum DeferredComponentInstallState {
+  /// The deferred component installation has been requested, but has not begun
+  /// download or installation yet.
+  requested,
+  /// The deferred component installation is awaiting further action.
+  pending,
+  /// User input is requried for continued installation.
+  requireUserConfirmation,
+  /// The deferred component is being downloaded.
+  downloading,
+  /// The deferred component has finished downloading but has not yet been
+  /// installed.
+  downloaded,
+  /// The downloaded deferred component is being installed.
+  installing,
+  /// The deferred component has previously been installed and all files/assets
+  /// are available, but has not been loaded into the VM in the current
+  /// app's session. Either `loadLibrary()` or [DeferredComponent.installDeferredComponent]
+  /// should be called to trigger the loading process to ensure the deferred component's
+  /// components can be safely used in the app.
+  installedPendingLoad,
+  /// The deferred component has been succesfully downloaded and installed. Assets and
+  /// code from the deferred component are ready to use.
+  installed,
+  /// An installation request has been asked to be canceled.
+  cancelling,
+  /// An installation request has been canceled.
+  canceled,
+  /// An installation request has failed.
+  failed,
+  /// The deferred component has not been requested nor installed or is unavailable.
+  unknown,
+}
+
+/// Manages the installation and loading of deferred component modules.
+///
```

Comment:
The docs above mention "Google Play Store". Is this an Android-only feature (for now)? May be worthwhile to mention that here as well?

Reply:
The initial version of this is Android only, but I do plan on extending this in some form to other platforms in the future.

Reply:
+1 We can update the docs when this does more on other platforms.